python/pyspark/ml/functions.py

```diff
@@ -106,6 +117,474 @@ def array_to_vector(col: Column) -> Column:
+        input_shapes: list[list[int] | None]
+        if isinstance(input_tensor_shapes, Mapping):
+            input_shapes = [None] * num_expected
+            for index, shape in input_tensor_shapes.items():
+                input_shapes[index] = shape
+        else:
+            input_shapes = input_tensor_shapes  # type: ignore
+
+        # iterate over partition, invoking predict_fn with ndarrays
+        for partition in data:
+            has_tuple = isinstance(partition, Tuple)  # type: ignore
+            has_tensors = _has_tensor_cols(partition)
+
+            # require input_tensor_shapes for any tensor columns
+            if has_tensors and not input_shapes:
+                raise ValueError("Tensor columns require input_tensor_shapes")
+
+            for batch in _batched(partition, batch_size):
+                num_actual = len(batch.columns)
+                if num_actual == num_expected and num_expected > 1:
+                    # input column per expected input, convert each column into param
+                    multi_inputs = [batch[col].to_numpy() for col in batch.columns]
+                    if input_shapes:
+                        if len(input_shapes) == num_actual:
+                            multi_inputs = [
+                                np.vstack(v).reshape([-1] + input_shapes[i])  # type: ignore
+                                if input_shapes[i]
+                                else v
+                                for i, v in enumerate(multi_inputs)
+                            ]
+                        else:
+                            raise ValueError("input_tensor_shapes must match columns")
+
+                    # run model prediction function on transformed (numpy) inputs
+                    preds = predict_fn(*multi_inputs)
+                elif num_expected == 1:
+                    # multiple input columns for single expected input
+                    if has_tensors:
+                        # tensor columns
+                        if len(batch.columns) == 1:
+                            # one tensor column and one expected input, vstack rows
+                            single_input = np.vstack(batch.iloc[:, 0])
+                        else:
+                            raise ValueError(
+                                "Multiple input columns found, but model expected a single "
+                                "input, use `struct` or `array` to combine columns into tensors."
+                            )
+                    else:
+                        # scalar columns
+                        if len(batch.columns) == 1:
+                            # single scalar column, remove extra dim
+                            single_input = np.squeeze(batch.to_numpy())
+                        elif not has_tuple:
+                            # columns grouped via struct/array, convert to single tensor
+                            single_input = batch.to_numpy()
```

Comment:
For this case, I think we should force input_shapes[0] to be `[len(batch.columns)]` ?

Reply:
similar to above.

Reply:
reply ditto.